[SPARK-34863][SQL] Support complex types for Parquet vectorized reader

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR adds support for complex types (e.g., list, map, array) for Spark's vectorized Parquet reader. In particular, this introduces the following changes:
1. Added a new class `ParquetColumnVector` which encapsulates all the necessary information needed when reading a Parquet column, including the `ParquetColumn` for the Parquet column, the repetition & definition levels (only allocated for a leaf-node of a complex type), as well as the reader for the column. In addition, it also contains logic for assembling nested columnar batches, via interpreting Parquet repetition & definition levels. 
2. Changes are made in `VectorizedParquetRecordReader` to initialize a list of `ParquetColumnVector` for the columns read.
3. `VectorizedColumnReader` now also creates a reader for repetition column. Depending on whether maximum repetition level is 0, the batch read is now split into two code paths, e.g., `readBatch` versus `readBatchNested`.
4. Added logic to handle complex type in `VectorizedRleValuesReader`. For data types involving only struct or primitive types, it still goes with the old `readBatch` method which now also saves definition levels into a vector for later assembly. Otherwise, for data types involving array or map, a separate code path `readBatchNested` is introduced to handle repetition levels.
This PR also introduced a new flag `spark.sql.parquet.enableNestedColumnVectorizedReader` which turns the feature on or off. By default it is on to facilitates all the Parquet related test coverage.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Whenever read schema containing complex types, at the moment Spark will fallback to the row-based reader in parquet-mr, which is much slower. As benchmark shows, by adding support into the vectorized reader, we can get ~15x on average speed up on reading struct fields, and ~1.5x when reading array of struct and map.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

With the PR Spark should now support reading complex types in its vectorized Parquet reader. A new config `spark.sql.parquet.enableNestedColumnVectorizedReader` is introduced to turn the feature on or off.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Added new unit tests.

sunchao committed:
wip

sunchao committed:
more tests

sunchao committed:
update golden files

sunchao committed:
more fixes

SparkQA commented:
**[Test build #145429 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145429/testReport)** for PR 34659 at commit [`10c3a45`](https://github.com/apache/spark/commit/10c3a450af62b0028601c862005a566d623af253).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49902/

cloud-fan commented:
cc @sadikovi

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49902/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49902/

sadikovi commented:
ACK. I will review it tomorrow or next week if that's okay 🙂.

SparkQA commented:
**[Test build #145429 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145429/testReport)** for PR 34659 at commit [`10c3a45`](https://github.com/apache/spark/commit/10c3a450af62b0028601c862005a566d623af253).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145429/

sunchao commented:
Thanks @sadikovi! take your time on this one :)

sadikovi commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -897,6 +897,14 @@ object SQLConf {
+  val PARQUET_VECTORIZED_READER_NESTED_COLUMN_ENABLED =
+    buildConf("spark.sql.parquet.enableNestedColumnVectorizedReader")
+      .doc("Enables vectorized Parquet decoding for nested columns (e.g., struct, list, map). " +
+          s"Note to enable this ${PARQUET_VECTORIZED_READER_ENABLED.key} also needs to be enabled.")
```
I am not sure what the guidelines are on a conf text but maybe this would sound a bit crisper: 

`Enables vectorized Parquet decoding for nested columns (e.g., struct, list, map). Requires ${PARQUET_VECTORIZED_READER_ENABLED.key} to be enabled.`

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,321 @@
+
+  void setColumnReader(VectorizedColumnReader reader) {
+    if (!isPrimitive) {
+      throw new IllegalStateException("can't set reader for non-primitive column");
```
nit: `Can't set...`

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,321 @@
+    }
+  }
+
+  void reset() {
```
I understand it requires a lot of typing but it would be good if we could add a small javadoc here and below to highlight that the methods do.

sunchao replied:
Sure, added.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,321 @@
+        vector.putArray(rowId, offset, 0);
+        rowId++;
+      } else if (definitionLevel > maxDefinitionLevel) {
+        // collection is defined and non-empty: find out how many top element there is till the
```
nit: find out how many elements are there until the start of the next array. Alternatively, find out the number of elements until the start of the next array.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,321 @@
+    return idx;
+  }
+
+  private int getCollectionSize(int maxRepetitionLevel, int idx) {
```
We might need javadoc for this and a couple of methods above to go along with the inline comments.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,321 @@
+        // definition levels: [2, 2, 2, 2, 2, 2, 2, 2]
+        //
+        // when calculating collection size for the outer array, we should only count repetition
+        // levels whose value is <= 1 (which is the max repetition level for the inner array)
```
Could you elaborate on this comment a bit more? Is it about the example above or in general?

sunchao replied:
You mean `<= 1` in the sentence? it is about the above example.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,321 @@
+  }
+
+  private int getCollectionSize(int maxRepetitionLevel, int idx) {
+    int size = 1;
```
I think `size = 1` requires a comment. Is it due to the fact that the collection is non-empty and we automatically add the first element? 

Depending on the idx, we could double count size or skip elements. I would propose to return whatever size was calculated based on index and add +1 or +K in the caller's code if required.

sunchao replied:
Yes, it is because the collection is non-empty, and `idx` is pointing to the next element after the first element in the collection. 

> Depending on the idx, we could double count size or skip elements

Hmm why this will happen? I think it's better to calculate the final size in the method instead of having to add +1 in every call site.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetReadState.java:
```diff
@@ -42,24 +44,52 @@
+
+  /* The following fields are only used when reading repeated values */
+
+  /** When processing repeated values, whether we've found the beginning of the first list after the
```
Javadoc should be updated.

sunchao replied:
Update what?

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetReadState.java:
```diff
@@ -42,24 +44,52 @@
+  /** When processing repeated types, the number of accumulated definition levels to process */
+  int numBatchedDefLevels;
+
+  /** When processing repeated types, whether we should skip the current batch of definition
```
Same here.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -39,12 +42,13 @@
 import org.apache.spark.sql.vectorized.ColumnarBatch;
 import org.apache.spark.sql.types.StructField;
 import org.apache.spark.sql.types.StructType;
+import scala.collection.JavaConverters;
```
Is this import required?

sunchao replied:
Yes it's used below in `checkColumn` method for example

sadikovi replied:
That should probably be moved after java imports 🙂. I am probably nit-picking here but maybe it would be a good nit to fix - it is up to you.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -303,55 +313,88 @@ public boolean nextBatch() throws IOException {
-      if (!t.isPrimitive() || t.isRepetition(Type.Repetition.REPEATED)) {
-        throw new UnsupportedOperationException("Complex types not supported.");
-      }
+    missingColumns = new HashSet<>();
```
I don't know if Spark runs java lint. Is it allowed to use `<>` with generic types?

sunchao replied:
Yes it's it allowed. This use pattern is pretty standard in Java (in fact IntelliJ will warn you if you put the generic type there).

sadikovi replied:
All good, I remember Java 6 used to complain if you use `<>` diamond type.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -303,55 +313,88 @@ public boolean nextBatch() throws IOException {
+   * conforms to the type of the file schema.
+   */
+  private void checkColumn(ParquetColumn column) throws IOException {
+    String[] path = JavaConverters.seqAsJavaList(column.path()).toArray(new String[0]);
```
Maybe the original path could be returned as an array...

sunchao replied:
It could .. although I'd have to change a lot of places in `ParquetSchemaSuite`. Since this is a relatively small issue, maybe let me address it separately?

sadikovi replied:
Yes, let's address separately.

sadikovi reviewed (commented):
Thanks for opening a PR. Looks very good but I need to re-review a couple of files. Tests look good, I remember there were a couple of tricky examples with array + struct + null nesting, let me find them and check, maybe we could add them to the tests.

On a more style-like note: could you make all of the comments to start with an uppercase letter? I find it easier to read them this way as they indicate the beginning of a paragraph. Thank you.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java:
```diff
@@ -176,8 +176,7 @@ public void initBatch(
           OnHeapColumnVector missingCol = new OnHeapColumnVector(capacity, dt);
-          missingCol.putNulls(0, capacity);
-          missingCol.setIsConstant();
+          missingCol.setAllNull();
```
Oh, almost forgot to ask. 

Is `missingCol.setAllNull();` equivalent to 
```
missingCol.putNulls(0, capacity);
missingCol.setIsConstant();
```
? I thought `setAllNull` was an alias for the two method calls but it does not appear to be the case.

sunchao replied:
Not quite the same. `setIsConstant` requires to allocate space for the constant elements, but `setAllNull` doesn't. This is useful for those missing columns that are of nested types.

dongjoon-hyun replied:
Is this related this PR or new test coverage?
This change looks like relying on the existing ORC tests only.
In that case, shall we make a separate PR for this ORC change?

sadikovi commented:
@sunchao Let me know if you would like me to revisit the code in this PR. Thanks.

sunchao commented:
Thanks @sadikovi , really appreciate your feedback! will address the comments soon! and then you can re-visit the PR.

agrawaldevesh commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -227,30 +230,340 @@ private void readBatchInternal(
         switch (mode) {
           case RLE:
```
I am not too familiar with this code, but where do we assert that these are the only two encoding choices. There is no default arm in this case.

sunchao replied:
The `MODE` enum is private to this class and the `mode` is initialized in this class, so I guess it should be fine. This maps to the Parquet spec for RLE/BitPacked encoding.

agrawaldevesh reviewed (commented):
Thanks: This would be a great speed up for Uber given its heavy reliance on Nested data. I am curious how often are nested fields selected in the resultSchema given nested column pruning in effect ?

Would it be possible to retain the benchmark in the older version of this PR https://github.com/apache/spark/pull/33695 and perhaps also add a case where some of the struct column values are NULL. 

How do we confirm that this hasn't slowed down the non nested code path ?

sunchao commented:
Thanks @agrawaldevesh . Will address your comments soon.

> How do we confirm that this hasn't slowed down the non nested code path ?

I've run the `DataSourceReadBenchmark` through GitHub workflow with and without the PR, and put the results [here](https://gist.github.com/sunchao/fe67e6846b9fc2a46412b6de7e89afad). It seems the results don't change much but there are some variances.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -227,30 +230,340 @@ private void readBatchInternal(
+
+                leftInBatch -= n;
+                if (n > 0) {
+                  repLevels.appendInts(n, 0);
```
Hmm what do you mean? we need to read in repetition levels in this method.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -227,30 +230,340 @@ private void readBatchInternal(
+      values, nulls, valuesReused, valueReader, updater);
+
+    while ((leftInBatch > 0 || !state.lastListCompleted) && leftInPage > 0) {
+      if (currentCount == 0 && !readNextGroup()) break;
```
Yea good point, let me change the non nested code, although I think for that case `readNextGroup` will never return false since we know exactly how many elements to read beforehand.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -227,30 +230,340 @@ private void readBatchInternal(
+                defLevelProcessor.skipValues(n);
+                rowId += n;
+                currentCount -= n;
+                leftInPage -= n;
```
it shouldn't since we don't add anything to the batch here (we are skipping rows).

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -227,30 +230,340 @@ private void readBatchInternal(
+                  defLevelProcessor.readValues(n);
+                }
+
+                rowId += n;
```
Yes better to move it here. The member vars (e.g., `currentCount`) tracks the state of the RLE/BitPacked block being read, while the local vars track the state of the input vectors.

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -897,6 +897,14 @@ object SQLConf {
+          s"Note to enable this ${PARQUET_VECTORIZED_READER_ENABLED.key} also needs to be enabled.")
+      .version("3.3.0")
+      .booleanConf
+      .createWithDefault(true)
```
This is to make sure we can pass all Parquet related tests :) I plan to turn it off later separately.

viirya replied:
Make sure we remember to turn it off as default before merging. 😄

sunchao committed:
address comments

SparkQA commented:
**[Test build #145831 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145831/testReport)** for PR 34659 at commit [`7b6e083`](https://github.com/apache/spark/commit/7b6e083833c807e44b76bc321d32e9c5857e648f).

sunchao commented:
Thanks @sadikovi and @agrawaldevesh ! I've addressed most of your comments, while some left to be answered. Please let me know what you think.

> On a more style-like note: could you make all of the comments to start with an uppercase letter? I find it easier to read them this way as they indicate the beginning of a paragraph.

Sure, it's done.

>  am curious how often are nested fields selected in the resultSchema given nested column pruning in effect ?

This is orthogonal I think. With or without nested column pruning users may still want to define nested schemas and efficiently read data of them. From my experience I've seen struct type being commonly used in data schemas and this PR should help a lot.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50306/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50306/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50306/

SparkQA commented:
**[Test build #145831 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145831/testReport)** for PR 34659 at commit [`7b6e083`](https://github.com/apache/spark/commit/7b6e083833c807e44b76bc321d32e9c5857e648f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145831/

cloud-fan commented:
@sadikovi do you have more comments?

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -303,55 +313,88 @@ public boolean nextBatch() throws IOException {
-          // Column is missing in data but the required data is non-nullable. This file is invalid.
-          throw new IOException("Required column is missing in data file. Col: " +
-            Arrays.toString(colPath));
+        for (ParquetColumn childColumn : JavaConverters.seqAsJavaList(column.children())) {
```
Same here, I think.

sunchao replied:
Thanks, yea I can raise another PR to use list in `ParquetColumn`.

sadikovi reviewed (approved):
LGTM, thanks for making the changes!

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java:
```diff
@@ -350,6 +355,14 @@ public WritableColumnVector reserveDictionaryIds(int capacity) {
+   *
+   * NOTE: this MUST be called after new elements are appended to child vectors of a struct vector.
+   */
+  public abstract void putStruct(int rowId, int offset);
```
I want to point out that this is a behavior change in `WritableColumnVector` so anyone who's using this to write struct will need to adopt to the new API. 

Wanna see if you have any concern on this @cloud-fan @sadikovi .

sadikovi replied:
IMHO, it is fine as the PR goes to master, this is supposed to be internal API anyway unless I am mistaken 🙂.

cloud-fan replied:
yea `WritableColumnVector` is an internal API

sunchao replied:
cool, thanks for confirming!

bersprockets commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:
```diff
@@ -169,8 +170,8 @@ class ParquetFileFormat
     conf.parquetVectorizedReaderEnabled && conf.wholeStageEnabled &&
-      schema.length <= conf.wholeStageMaxNumFields &&
-      schema.forall(_.dataType.isInstanceOf[AtomicType])
+      ParquetFileFormat.isBatchReadSupported(conf, schema) &&
```
When ```spark.sql.parquet.enableNestedColumnVectorizedReader=false```, which will be the normal case initially, I think returningBatch will always be false, even for non-complex cases, since this [line](https://github.com/apache/spark/blob/7b6e083833c807e44b76bc321d32e9c5857e648f/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala#L249) always passes a struct (resultSchema) to ```supportBatch```.

At least, that's what my testing seems to show. With this PR and ```spark.sql.parquet.enableNestedColumnVectorizedReader=false```, returningBatch is false for a non-complex case. But on master, returningBatch is true for the same test:

```
spark.range(0, 10).map { x => (x, x + 1, x + 3) }.toDF("a", "b", "c").
  write.mode("overwrite").format("parquet").save("simple_parquet")

sql("select a, b from `parquet`.`simple_parquet`").collect
```
and...
```
bash-3.2$ git diff | cat
diff --git a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
index bd8d11d827..96182aafa9 100644
--- a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
+++ b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
@@ -340,6 +340,7 @@ class ParquetFileFormat
           vectorizedReader.initialize(split, hadoopAttemptContext)
           logDebug(s"Appending $partitionSchema ${file.partitionValues}")
           vectorizedReader.initBatch(partitionSchema, file.partitionValues)
+          print(s"returningBatch is $returningBatch\n")
           if (returningBatch) {
             vectorizedReader.enableReturningBatches()
           }
bash-3.2$ 
```

sunchao replied:
Good catch! I think it should be:
```scala
schema.forall(f => ParquetFileFormat.isBatchReadSupported(conf, f.dataType))
```

sunchao committed:
fix logic to enable/disable batch reads

SparkQA commented:
**[Test build #146411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146411/testReport)** for PR 34659 at commit [`aff8504`](https://github.com/apache/spark/commit/aff8504d4130095ee1ecfdab3f3d2d9110d40400).

SparkQA commented:
**[Test build #146411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146411/testReport)** for PR 34659 at commit [`aff8504`](https://github.com/apache/spark/commit/aff8504d4130095ee1ecfdab3f3d2d9110d40400).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146411/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50886/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50886/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50886/

sunchao committed:
fix lint

SparkQA commented:
**[Test build #146490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146490/testReport)** for PR 34659 at commit [`77b2aec`](https://github.com/apache/spark/commit/77b2aec91350244dcf02489c2bc6f0d09f7bab9d).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50965/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50965/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50965/

SparkQA commented:
**[Test build #146490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146490/testReport)** for PR 34659 at commit [`77b2aec`](https://github.com/apache/spark/commit/77b2aec91350244dcf02489c2bc6f0d09f7bab9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146490/

bersprockets commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java:
```diff
@@ -103,9 +104,14 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
-    this.sparkSchema = StructType$.MODULE$.fromString(sparkRequestedSchemaString);
+    StructType sparkRequestedSchema = StructType$.MODULE$.fromString(sparkRequestedSchemaString);
+    ParquetToSparkSchemaConverter converter = new ParquetToSparkSchemaConverter(configuration);
+    this.parquetColumn = converter.convertParquetColumn(requestedSchema,
```
This line has a performance impact on parquet files with very wide records.

Take, for example, with a parquet dataset with:

- 6000 bigint columns
- 25 files
- 5000 records per file (total of 125000 records)

Reading all the records in this parquet dataset with a single executor thread is 42% slower with this PR than with the baseline (commit f361ad8d2e on the master branch).

The main culprit is the n**2 issue [here](https://github.com/apache/spark/blob/f361ad8d2eb067ec86483a294771f1976f31a981/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L95)  in ParquetToSparkSchemaConverter#convertInternal, which was not exercised anywhere except by unit tests until your PR (as far as I can tell).

By the way, even with a patch for the above n**2 issue, this PR is still 10% slower than the baseline when testing the wide row case. I haven't confirmed it, but I think this is due to the extra OnHeapColumnVector instances created in ParquetColumnVector, for which you already have a TODO.

sunchao replied:
Thanks! let me try to reproduce the above scenario (it'd be great if you already have some code snippet to share) locally and see what's the causing the perf issue in `ParquetToSparkSchemaConverter#convertInternal`. Ideally we can skip the code path if the schema doesn't contain complex types.

bersprockets replied:
>see what's the causing the perf issue in ParquetToSparkSchemaConverter#convertInternal

It's definitely the n\*n (or n\*(n/2) on average) loop here:
```
    val converted = (0 until groupColumn.getChildrenCount).map { i =>
      val field = groupColumn.getChild(i)
      val fieldFromReadSchema = sparkReadSchema.flatMap { schema =>
        schema.find(f => isSameFieldName(f.name, field.getName, caseSensitive))
      }
```

`schema.find` should instead be a map lookup, or something like that.

>it'd be great if you already have some code snippet to share

Sure, in a verbose state:

```
// test for wide rows

val home_candidate = sys.env("HOME")
val home = if (!home_candidate.endsWith("/")) {
  s"${home_candidate}/"
} else {
  home_candidate
}

val width = 6000

val selectExpr = (1 to width).map { i =>
  s"value as c$i"
}


import scala.util.Random
val r = new Random(65657652L)

val loopCount = 25
for (i <- 0 until loopCount) {
  // println(s"iteration $i")
  val df = spark.range(5000).map(_ => r.nextLong).toDF().selectExpr(selectExpr: _*)
  val mode = if (i == 0) {
    "overwrite"
  } else {
    "append"
  }
  df.coalesce(1).write.mode(mode).format("parquet").save(s"${home}data/wide_row_parquet")
}


// read test

val home_candidate = sys.env("HOME")
val home = if (!home_candidate.endsWith("/")) {
  s"${home_candidate}/"
} else {
  home_candidate
}

spark.read.parquet(s"${home}data/wide_row_parquet").createOrReplaceTempView("df")
val startTime = System.currentTimeMillis
sql("select * from df where (c5*c12) == 12").collect
(System.currentTimeMillis - startTime)/60.0/1000
```

sunchao replied:
Thanks again @bersprockets ! let me address the issue and update the PR.

sunchao replied:
Updated the PR. I also re-run the above test which used to take 1.05m before the fix, and improved to 0.58m afterwards, which is very close to the number in master branch. 

@bersprockets it'd be great if you can verify it too. Thanks.

bersprockets replied:
Yes, you've eliminated the schema lookup penalty.

There still seems to be the remaining 10-15% performance penalty. Earlier, I speculated on the cause of this particular penalty:

>this PR is still 10% slower than the baseline when testing the wide row case. I haven't confirmed it, but I think this is due to the extra OnHeapColumnVector instances created in ParquetColumnVector, for which you already have a TODO.

So maybe that will get resolved once the TODO is resolved.

sunchao replied:
Thanks, yes I'm aware of the other one: we should not need to allocate the vectors for definition & repetition levels when the schema is flat. I'm hoping to address this separately with another PR though - don't want to make this one too bloated :)

BTW @bersprockets : how can I reproduce the 10-15% performance penalty? I was using the above code snippet and got almost the same numbers on my machine with the latest fix.

bersprockets replied:
>how can I reproduce the 10-15% performance penalty

Maybe my old 2015 2.5 GHz MBP just feels it more strongly.

I launched spark-shell it with this command line
```
bin/spark-shell --driver-memory 5g --master "local[1]"
```
Maybe the 5g makes a difference (e.g., more pressure on the garbage collector when there are lots of OnHeapColumnVector instances).

sunchao replied:
I have the follow-up to skip creating these column vectors and reading definition levels if unnecessary done [here](https://github.com/sunchao/spark/commit/f968791635db1d679638511fc4ff01d2b1eccff0). I can include it to the PR if it doesn't complicated the review :) cc @sadikovi

sunchao committed:
Address perf issue in ParquetSchemaConverter

klaus-xiong commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala:
```diff
@@ -169,8 +170,8 @@ class ParquetFileFormat
-      schema.length <= conf.wholeStageMaxNumFields &&
-      schema.forall(_.dataType.isInstanceOf[AtomicType])
+      ParquetUtils.isBatchReadSupportedForSchema(conf, schema) &&
+        !WholeStageCodegenExec.isTooManyFields(conf, schema)
```
indent

klaus-xiong reviewed (commented):
This work is great :) .

LuciferYang commented:
@sunchao Is this pr still going on?

sunchao commented:
@LuciferYang yes it is, let me rebase it to the latest master and address the comments by @klaus-xiong above.

sunchao committed:
space

sunchao commented:
Could you review this? @cloud-fan @dongjoon-hyun @viirya ? It'd be great if this can catch Spark 3.3 release.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileBasedDataSourceTest.scala:
```diff
@@ -52,7 +54,8 @@ private[sql] trait FileBasedDataSourceTest extends SQLTestUtils {
     if (testVectorized) {
-      withSQLConf(vectorizedReaderEnabledKey -> "true") {
+      withSQLConf(vectorizedReaderEnabledKey -> "true",
+          vectorizedReaderNestedEnabledKey -> "true") {
```
Could you revise the test framework to have a complete test coverage for both `true` and `false`, @sunchao ?

sunchao replied:
Sure, good idea.

LuciferYang commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,359 @@
+
+    this.column = column;
+    this.vector = vector;
+    this.children = new ArrayList<>();
```
Should we lazy initializing `children` after line 86 and use `column.children()` to specify the `ArrayList` size

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:
```diff
@@ -144,6 +145,30 @@ object ParquetUtils {
+  /**
+   * Whether columnar read is supported for the input `schema`.
+   */
+  def isBatchReadSupportedForSchema(sqlConf: SQLConf, schema: StructType): Boolean =
```
It would be great to design new functions minimally which receive the minimal information instead of passing over the full `SQLConf` recursively. For example, could you revise
- `isBatchReadSupportedForSchema` by handing over the values of `sqlConf.parquetVectorizedReaderEnabled` and `sqlConf.parquetVectorizedReaderNestedColumnEnabled`
- `isBatchReadSupported` by handing over the value of `sqlConf.parquetVectorizedReaderNestedColumnEnabled` from `isBatchReadSupportedForSchema`.

directly instead of `SQLConf` recursively?

sunchao replied:
Makes sense. Let me change it.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:
```diff
@@ -144,6 +145,30 @@ object ParquetUtils {
+    sqlConf.parquetVectorizedReaderEnabled &&
+      schema.forall(f => isBatchReadSupported(sqlConf, f.dataType))
+
+  def isBatchReadSupported(sqlConf: SQLConf, dt: DataType): Boolean = dt match {
```
This looks like visible for testing only. If you don't mind, could you make this `private` and rewrite the only usage in `ParquetFileFormatSuite` with `invokePrivate`?
```
sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormatSuite.scala:          assert(ParquetUtils.isBatchReadSupported(conf, dt) == expected)
```

sunchao replied:
Sure

dongjoon-hyun reviewed (dismissed):
Hi, @sunchao . I made a few comments to make this PR move forward.
- Missing test coverage: https://github.com/apache/spark/pull/34659/files#r803324113
- New function design: https://github.com/apache/spark/pull/34659/files#r803330689
- Function visibility: https://github.com/apache/spark/pull/34659/files#r803331855
- Spinning off ORC code change **after** this PR: https://github.com/apache/spark/pull/34659/files#r803334124
- Lastly, let's spin off (non-Parquet) `ColumnVector` layer as an independent PR **before** this PR.
```
 sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java               |  2 +-
 sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OffHeapColumnVector.java  | 25 +++++++++++++++++++++++++
 sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java   | 28 +++++++++++++++++++++++++++-
 sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java | 82 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------
 sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala  |  4 ++++
 sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala | 15 +++++++++++++++
```

sunchao committed:
address comments

dongjoon-hyun commented:
Thanks!

sunchao committed:
remove more unnecessary changes

sadikovi commented:
Let me take another look, it seems there are have been quite a few changes since 🙂 .

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+    this.children = new ArrayList<>();
+    this.isPrimitive = column.isPrimitive();
+
+    if (missingColumns.contains(column)) {
```
I would like to confirm that this check is by reference, e.g. if there are two instances of ParquetColumn with the same name and only one is in the set, the other one would return `false` when checking. If that is expected, it is all good.

sunchao replied:
Yes, this does compare by reference. `parquetColumn` in `VectorizedParquetRecordReader` is created in its constructor, and then later on it is used in both constructing the `missingColumns` field and `columnVectors` field.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +213,358 @@ private void readBatchInternal(
+
+    void readValues(int n) {
+      if (!state.shouldSkip) {
+        state.numBatchedDefLevels += n;
```
Apologies for a potentially dumb question, but why don't we call `reader.readValues(...)` here?

sunchao replied:
This is to enable batch reading on the definition levels & values as much as possible. For instance, if we have the following triplets:

| repetition level | definition level | values |
----------------|---------------|--------
| 3 | 1 | 0 |
| 2 | 1 | 1 |
| 3 | 1 | 2 |
| 2 | 1 | 3 |
| 3 | 1 | 4 |
| 2 | 1 | 5 |

First we'd have to read the repetition levels, which must be encoded with bit-packing and therefore have to be read one by one. On the other hand, the definition levels are probably encoded in RLE and therefore, together with the values, can be read in one batch.

If we don't use this, then we'd have to call `reader.readValues` for each repetition level read here, which is more expensive I think.

sadikovi commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java:
```diff
@@ -763,6 +805,10 @@ protected boolean isArray() {
   }
 
+  protected boolean isStruct() {
+    return type instanceof StructType || type instanceof CalendarIntervalType;
```
Do you know if we have tests for this? I am curious what happens if someone adds a union type or another composite type, will we need to update this method?

sunchao replied:
Hmm this is no longer used - I created it for the `WritableColumnVector` change I think, but now the change is no longer needed. Let me remove this.

sadikovi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:
```diff
@@ -187,6 +188,30 @@ object ParquetUtils {
+      sqlConf.parquetVectorizedReaderNestedColumnEnabled &&
+          isBatchReadSupported(sqlConf, mt.keyType) &&
+          isBatchReadSupported(sqlConf, mt.valueType)
+    case st: StructType =>
```
Should we support CalendarIntervalType here as well? I saw a method checking on that type also.

sunchao replied:
I'm not sure if we even support reading & writing of `CalendarIntervalType` in Parquet at the moment - it seems we don't from looking at `ParquetWriteSupport.makeWriter`. IIUC it is also going to be deprecated by `YearMonthIntervalType` and `DayTimeIntervalType`, is that correct @MaxGekk ?

sadikovi reviewed (approved):
LGTM. Looks great! Thanks for working on it.
I left a few questions, mostly informational. Let me know if something is unclear.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.base.Preconditions;
+import org.apache.spark.memory.MemoryMode;
```
```suggestion
import com.google.common.base.Preconditions;

import org.apache.spark.memory.MemoryMode;
```

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java:
```diff
@@ -738,6 +774,12 @@ public WritableColumnVector arrayData() {
   protected boolean isConstant;
 
+  /**
+   * True if this column only contains nulls. Comparing to 'isConstant' above, this doesn't require
+   * any allocation of space.
```
```suggestion
   * True if this column only contains nulls. This means the column values never change, even
   * across resets. Comparing to 'isConstant' above, this doesn't require
   * any allocation of space.
```

Based on `reset` change.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +213,358 @@ private void readBatchInternal(
             break;
         }
-        offset += n;
+        state.levelOffset += n;
```
why don't update `state.levelOffset` in place like `valueOffset`? Could be more consistent.

sunchao replied:
This is because `valueOffset` is only updated when the value is not null, but `levelOffset` is updated in all the cases. I could move this to the end of each `case` clause above, but I feel it's not too helpful.

viirya reviewed (commented):
So far looks good. I will continue reviewing this change this afternoon and tomorrow.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +213,358 @@ private void readBatchInternal(
+    state.rowId = rowId;
+  }
+
+  public void readBatchNested(
```
Seems this is used to read repetition level column only? Could you add a comment here too?

sunchao replied:
Yes this is only for repeated values, e.g., array, map. I can rename it to `readBatchRepeated` if that's more clearer.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +213,358 @@ private void readBatchInternal(
+            if (!state.shouldSkip) {
+              repLevels.appendInts(valuesLeftInBlock, currentValue);
+            }
+            state.numBatchedDefLevels += valuesLeftInBlock;
```
Don't we need to read definition level values `valuesLeftInBlock`?

sunchao replied:
Yes, the definition levels and values are read in `DefLevelProcessor`, in a batched fashion.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+    //     ==> value is null
+    //   3. definitionLevel < maxDefinitionLevel - 1
+    //     ==> value doesn't exist since one of its optional parents is null
+    //   4. definitionLevel > maxDefinitionLevel
+    //     ==> value is a nested element within an array or map
```
Any example for case 4? You mean the `maxDefinitionLevel` is the definition level of the array? Then isn't the `maxDefinitionLevel`  should be `array's definitionLevel + 1`

sunchao replied:
For example, an optional array with optional element is represented by:
```
                                        RL   DL
message schema {                        0    0
  optional group _1 (LIST) {            0    1
    repeated group list {               1    2
      optional binary element (UTF8);   1    3
    }
  }
}
```

here `RL` and `DL` are the max repetition and definition level for the schema element at the corresponding positions. The `maxDefinitionLevel` here is the max definition level of the array itself, which is 1. Any array element should have definition level > 1.

viirya replied:
Oh, I see. `maxDefinitionLevel` is confused me that as I thought it is max def level for elements. Because this equation `definitionLevel > maxDefinitionLevel` looks not reasonable. Actually you indicate two def levels at different element (array and nested element).

sunchao replied:
Yea `maxDefinitionLevel` and `maxRepetitionLevel` is for the collection itself here. For collection elements, the naming would be `maxElementDefinitionLevel` and `maxElementRepetitionLevel`.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+  private int getNextCollectionStart(int maxRepetitionLevel, int idx) {
+    idx += 1;
+    for (; idx < repetitionLevels.getElementsAppended(); idx++) {
+      if (repetitionLevels.getInt(idx) <= maxRepetitionLevel) {
```
Isn't `repetitionLevels.getInt(idx) < maxRepetitionLevel`?

viirya replied:
If it is `maxRepetitionLevel`, `idx` is still repeated at the same array, no?

sunchao replied:
The `maxRepetitionLevel` is the repetition level of the array itself. Drawing the previous example:
```
                                        RL   DL
message schema {                        0    0
  optional group _1 (LIST) {            0    1
    repeated group list {               1    2
      optional binary element (UTF8);   1    3
    }
  }
}
```

the `maxRepetitionLevel` is 0 in this case. Therefore, for an array of `[[0, 1, 2], [3, 4, 5]]`, it's repetition levels will be: `[0, 1, 1, 0, 1, 1]`. The if condition returns true when `repetitionLevels.getInt(idx)` is equal to 0.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+
+    ParquetColumn child = column.children().apply(0);
+    int maxElementDefinitionLevel = child.definitionLevel();
+    if (!child.required()) maxElementDefinitionLevel -= 1;
```
Hmm, isn't that we should do `maxElementDefinitionLevel -= 1;` when `child.required()`?

sunchao replied:
This is because `child.definitionLevel()` already incremented when the child is optional, so we need to negate it here. Still considering the example above. For optional elements:

```
                                        RL   DL
message schema {                        0    0
  optional group _1 (LIST) {            0    1
    repeated group list {               1    2
      optional binary element (UTF8);   1    3
    }
  }
}
```
The goal is to decide whether a value is an array element or not. Here `maxElementDefinitionLevel` is 3 - 1 = 2, and anything < 2 is not an array element.

```
```
                                        RL   DL
message schema {                        0    0
  optional group _1 (LIST) {            0    1
    repeated group list {               1    2
      required binary element (UTF8);   1    2
    }
  }
}
```
Similarly for required element, `maxElementDefinitionLevel` is 2.

I can change the variable name `maxElementDefinitionLevel` to something else if it's confusing :)

viirya replied:
> Similarly for required element, maxElementDefinitionLevel is 2.

Yea, this is why I'm confused here. Seems it is when `child.required()`, element's def level is 2, but in the code, `maxElementDefinitionLevel` is 2 when `!child.required()`.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+        //  length: 0 3
+        //  offset: 0 1
+        //
+        // It's important that for the second element `[a, b, c]`, the offset starts from 1
```
```suggestion
        // It's important that for the second element `[a, b, c]`, the offset in the array (not the elements) starts from 1
```

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+        // since otherwise we'd include the first null element from child array in the result.
+        offset += 1;
+      }
+      if (definitionLevel <= maxDefinitionLevel - 1) {
```
why this is not `else if`? Could two cases happen at the same time?

sunchao replied:
Yes we can merge them (just need to add `offset += 1` in the first two if branches. I separated it out mainly to add the block of comments explaining why we need to increment `offset`.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java:
```diff
@@ -707,15 +711,47 @@ public WritableColumnVector arrayData() {
+  /**
+   * Increment number of elements appended by 'num'.
+   *
+   * This is useful when one wants to the 'putXXX' API to add new elements to the vector, but
```
```suggestion
   * This is useful when one wants to use the 'putXXX' API to add new elements to the vector, but
```

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java:
```diff
@@ -0,0 +1,379 @@
+    vector.addElementsAppended(rowId);
+  }
+
+  private void calculateStructOffsets() {
```
nit: this seems not about offsets but adding nulls into struct and update row counts.

viirya reviewed (commented):
This already looks good. I've only a few minor comments and questions.

sunchao committed:
address more comments

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:
```diff
@@ -157,7 +160,11 @@ private boolean isLazyDecodingSupported(PrimitiveType.PrimitiveTypeName typeName
   /**
    * Reads `total` values from this columnReader into column.
```
Yes this is correct.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:
```diff
@@ -214,24 +231,32 @@ void readBatch(int total, WritableColumnVector column) throws IOException {
 
   private int readPage() {
     DataPage page = pageReader.readPage();
+    if (page == null) {
+      return -1;
+    }
```
Yes, I have some comments in line 182 above:

> we've read all the pages; this could happen when we're reading a repeated list and we don't know where the list will end until we've seen all the pages.

For primitive types, we know a) the exact number of total values to read, and 2) the total number of values for each page. Therefore we know which page is the last one.

However, a repeated list could span multiple pages in Parquet, and therefore we don't really know which page is the last, until the Parquet page returns null.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -246,25 +249,25 @@ private void initBatch(
+
+    columnVectors = new ParquetColumnVector[sparkSchema.fields().length];
+    for (int i = 0; i < columnVectors.length; i++) {
+      columnVectors[i] = new ParquetColumnVector(parquetColumn.children().apply(i),
```
The `parquetColumn` corresponds to the `StructType` for Spark's requested schema, so we need to apply on the children of it. Let me know if I understand your question correctly.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +213,358 @@ private void readBatchInternal(
+      valueReader, updater);
+  }
+
+  public void readIntegersNested(
```
This is used for decoding dictionary indices. Will add some comments.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +213,358 @@ private void readBatchInternal(
+  }
+
+  /**
+   * Keep reading repetition level values from the page until either: 1) we've read enough
```
No, it reads definition levels & values too. It's just the number of rep_level/def_level/value to read is controlled by repetition levels.

sunchao committed:
address more comments

sunchao commented:
Thanks @viirya for reviewing! I've addressed your comments, and please take another look.

viirya commented:
lgtm. I may merge this in next few days if no more comments. cc @cloud-fan @dongjoon-hyun

dongjoon-hyun reviewed (approved):
+1, LGTM. As @viirya  mentioned, it's time to disable the config. ;)
(For a record, the linter failure is irrelevant).

sunchao committed:
disable by default

dongjoon-hyun commented:
Hi, @sunchao . After rebasing and force-pushing, the PR branch is broken. Could you take a look?
```
[error] /home/runner/work/spark/spark/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:426:22: not enough arguments for constructor TaskContextImpl: (stageId: Int, stageAttemptNumber: Int, partitionId: Int, taskAttemptId: Long, attemptNumber: Int, numPartitions: Int, taskMemoryManager: org.apache.spark.memory.TaskMemoryManager, localProperties: java.util.Properties, metricsSystem: org.apache.spark.metrics.MetricsSystem, taskMetrics: org.apache.spark.executor.TaskMetrics, cpus: Int, resources: Map[String,org.apache.spark.resource.ResourceInformation])org.apache.spark.TaskContextImpl.
(https://github.com/sunchao/spark/runs/5778352461?check_suite_focus=true#step:9:600)
[error] Unspecified value parameter metricsSystem.
(https://github.com/sunchao/spark/runs/5778352461?check_suite_focus=true#step:9:601)
[error]       val context1 = new TaskContextImpl(0, 0, 0, 0, 0, null, null, null, cpus = 0)
(https://github.com/sunchao/spark/runs/5778352461?check_suite_focus=true#step:9:602)
[error]                      ^
(https://github.com/sunchao/spark/runs/5778352461?check_suite_focus=true#step:9:603)
[error] one error found
```

LuciferYang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:
```diff
@@ -187,6 +188,30 @@ object ParquetUtils {
+          isBatchReadSupported(sqlConf, mt.valueType)
+    case st: StructType =>
+      sqlConf.parquetVectorizedReaderNestedColumnEnabled &&
+          st.fields.forall(f => isBatchReadSupported(sqlConf, f.dataType))
```
nit: should be 2 spaces here

LuciferYang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:
```diff
@@ -187,6 +188,30 @@ object ParquetUtils {
+      true
+    case at: ArrayType =>
+      sqlConf.parquetVectorizedReaderNestedColumnEnabled &&
+          isBatchReadSupported(sqlConf, at.elementType)
```
nit: should be 2 spaces here

LuciferYang commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:
```diff
@@ -187,6 +188,30 @@ object ParquetUtils {
+          isBatchReadSupported(sqlConf, at.elementType)
+    case mt: MapType =>
+      sqlConf.parquetVectorizedReaderNestedColumnEnabled &&
+          isBatchReadSupported(sqlConf, mt.keyType) &&
```
ditto

sunchao replied:
let me check all the places with 4 space indentation ... I need to fix my IntelliJ settings.

LuciferYang commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +215,379 @@ private void readBatchInternal(
+            nulls.putNulls(state.valueOffset, num);
+            state.valueOffset += num;
+          }
+          defLevels.putInts(state.levelOffset, num, currentValue);
```
line 556 and 561 (`state.valueOffset += num;`) can move before line 563

sunchao replied:
👍 actually the comments here can also be removed

LuciferYang commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -210,48 +215,379 @@ private void readBatchInternal(
+          }
+          defLevels.putInts(state.levelOffset, num, currentValue);
+          break;
+        case PACKED:
```
line 565 ~ 575 looks same as line 247 ~ 257 except for the comments

sunchao replied:
Let me split this into a util method.

LuciferYang commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java:
```diff
@@ -714,15 +718,47 @@ public WritableColumnVector arrayData() {
+  /**
+   * Marks this column only contains null values.
+   */
+  public final void setAllNull() {
```
1. `column.numNulls()` may still 0 after we call  `setAllNull()`
2.  do `column.setAllNull()` and `column.putXXX`, then `column.isAllNull` still true.

The above two scenes look strange

sunchao replied:
This is similar to `setIsConstant` though: you can also call `column.putXXX` after the call.

LuciferYang replied:
OK, let's ignore this for the time being

sunchao committed:
address comments and turn on config

sunchao commented:
FYI I plan to open a follow-up PR to disable the config and also update related tests.

viirya commented:
Sounds good to me.

LuciferYang reviewed (approved):
LGTM

viirya commented:
Thanks all. Merging to master/3.3. @sunchao Please remember to open a follow-up PR to disable the config and update related tests before 3.3 release. Thanks.

sunchao commented:
Thanks all for the review!!! @viirya I just opened #36055 for the follow-up.
